external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaRelation.scala

```diff
@@ -102,17 +103,35 @@ private[kafka010] class KafkaRelation(
     // Create an RDD that reads from Kafka and get the (key, value) pair as byte arrays.
     val executorKafkaParams =
       KafkaSourceProvider.kafkaParamsForExecutors(specifiedKafkaParams, uniqueGroupId)
-    val rdd = new KafkaSourceRDD(
-      sqlContext.sparkContext, executorKafkaParams, offsetRanges,
-      pollTimeoutMs, failOnDataLoss, reuseKafkaConsumer = false).map { cr =>
-      InternalRow(
-        cr.key,
-        cr.value,
-        UTF8String.fromString(cr.topic),
-        cr.partition,
-        cr.offset,
-        DateTimeUtils.fromJavaTimestamp(new java.sql.Timestamp(cr.timestamp)),
-        cr.timestampType.id)
+    val rdd = if (includeHeaders) {
+      new KafkaSourceRDD(
+        sqlContext.sparkContext, executorKafkaParams, offsetRanges,
+        pollTimeoutMs, failOnDataLoss, reuseKafkaConsumer = false).map { cr =>
+          InternalRow(
```

Comment:
Same here. Rather than duplicate all this, construct a Seq of the values and optionally add the last one if headers are included. Then make an Internalrow from it.

Reply:
I first took that approach, but it results in a serialization error, dislike to `KafkaOffsetReader` above.